### ISSUE TITLE
Efficient pending struct locking

### ISSUE THREAD

**Issue by kvark** [kvark/froggy#44]
Currently wrapped in `Arc<Mutex<...>>`. It's designed to only be locked and changed for short periods of time. We might want to explore more efficient solutions than a `Mutex`, starting with `parking_lot` of #23.

I'm hoping we can get something custom, based on atomics.

**Comment by DaseinPhaos** [comment]
Some rough explorations:

`std::sync::Mutex` uses some boxed platform-dependent locks. (Namely, [`pthread_mutex` on unix](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/mutex.rs) and [`SRWLock` or `CriticalSection` on Windows](https://github.com/rust-lang/rust/tree/master/src/libstd/sys/windows).) Invocation to `std::sync::Mutex::lock()` essentially just invokes these platform-specific locking mechanisms (through yet another round of referencing due to `Box`), which may or may not be efficient.

Instead of a boxed platform-dependent mutex, [`parking_lot::Mutex` uses an `AtomicUsize`](https://github.com/Amanieu/parking_lot/blob/master/src/raw_mutex.rs). AFAICT, its `lock` method tries to spin for a few rounds before setting the `Mutex` into a "parking state". On Windows, switching to this crate yields up to 1.5x performance improvements. However, benchmark results on Linux(Travis) is not ideal.

Another crate, [`spin`](https://github.com/mvdnes/spin-rs) ([a "simple spinlock" based on an `AtomicBool`](https://github.com/mvdnes/spin-rs/blob/master/src/mutex.rs)), might also be helpful. Testing both on Windows and Linux show significant performance improvements.

Another concern IMO is that, currently our benchmarks can't fully reflect the performance of pending structs: we don't do a lot of `clone`ing on `Pointer`s. The only relevant operation is pointer dropping.

## Benchmarks from Travis

### with `std::sync::Mutex`

```text
test bench_build  ... bench:   1,477,181 ns/iter (+/- 269,511)
test bench_update ... bench:      30,363 ns/iter (+/- 451)
...
test bench_build  ... bench:   2,139,234 ns/iter (+/- 35,303)
test bench_update ... bench:      40,762 ns/iter (+/- 539)
...
test bench_build  ... bench:   1,277,839 ns/iter (+/- 60,088)
test bench_update ... bench:      11,664 ns/iter (+/- 146)
...
test bench_build  ... bench:   1,235,994 ns/iter (+/- 22,651)
test bench_update ... bench:      13,068 ns/iter (+/- 132)

```

## with `parking_lot::Mutex`

```text
test bench_build  ... bench:   1,236,221 ns/iter (+/- 214,091) # 84%
test bench_update ... bench:      30,291 ns/iter (+/- 5,222) # 99%
...
test bench_build  ... bench:   2,365,487 ns/iter (+/- 487,612) # 110%
test bench_update ... bench:      47,387 ns/iter (+/- 11,985) # 116 %
...
test bench_build  ... bench:   1,120,088 ns/iter (+/- 158,446) # 88%
test bench_update ... bench:      11,892 ns/iter (+/- 376) # 102%
...
test bench_build  ... bench:   1,019,715 ns/iter (+/- 293,243) # 82%
test bench_update ... bench:      13,189 ns/iter (+/- 7,090) # 101%
```

## with `spin::Mutex`

```text
test bench_build  ... bench:     992,106 ns/iter (+/- 8,076) # 67%
test bench_update ... bench:      29,749 ns/iter (+/- 316) # 98%
...
test bench_build  ... bench:   1,751,539 ns/iter (+/- 233,598) # 82%
test bench_update ... bench:      40,571 ns/iter (+/- 756) # 99%
...
test bench_build  ... bench:     777,955 ns/iter (+/- 86,125) # 61%
test bench_update ... bench:      11,730 ns/iter (+/- 1,140) # 101%
...
test bench_build  ... bench:     783,487 ns/iter (+/- 33,882) # 63%
test bench_update ... bench:      13,963 ns/iter (+/- 207) # 106%
```

**Comment by kvark** [comment]
@DaseinPhaos wow, this is some really nice research, thank you!

> Another concern IMO is that, currently our benchmarks can't fully reflect the performance of pending structs: we don't do a lot of cloneing on Pointers. The only relevant operation is pointer dropping.

Absolutely! It would be interesting to come up with some more dynamic tests, and maybe propagate them to `ecs_bench`.

> Another crate, spin (a "simple spinlock" based on an AtomicBool), might also be helpful. Testing both on Windows and Linux show significant performance improvements.

It's a clear winner according to the benchmark results you posted. Let's get it in, and keep looking for even better solutions?

## My vision

I think the optimal solution, whatever it is, includes the knowledge of different parts of `Pending`. E.g. if you need to add a reference (a very typical operation, done e.g. on pointer cloning), you don't care about either `epoch` or `sub_ref` fields, and whether they are used at the moment. Thus, I'd expect the best gains to come from using lock-free structures for `add_ref`, `sub_ref`, and `epoch`, without involving any mutexes at all. [crossbeam](https://crates.io/crates/crossbeam) anyone?

**Comment by DaseinPhaos** [comment]
@kvark A PR for `spin` then?

> Thus, I'd expect the best gains to come from using lock-free structures for add_ref, sub_ref, and epoch, without involving any mutexes at all. crossbeam anyone?

Sounds cool! Tried to switch `add_ref` and `sub_ref`s from `Vec` into data structures provided by `crossbeam` (since we need random accessing to `epoch`, I left it as a `Mutex<Vec<_>>`), but the benchmark result seems to get worse: using both `TreiberStack` and `MsQueue` have two times longer `build` time, `SegQueue` is a bit better, yet still worse than the current master. 
I guess under our current bench routine the node based nature of these structs outweighs their lock-free promises. I don't know much (if any at all) about lock-free data structures though, so maybe the fault was on me. Help wanted!

**Comment by kvark** [comment]
Thanks for testing `crossbeam` structs! Once again, excellent work :)
Yes, I suggest we go with `spen` for the time being, and start thinking on how to improve our tests.

**Comment by kvark** [comment]
Considering this is fixed by #46
